packages/teams-js/src/public/monetization.ts

```diff
@@ -39,12 +54,36 @@ export namespace monetization {
    *
    * @internal
    */
-  export function openPurchaseExperience(callback: (error: SdkError | null) => void, planInfo?: PlanInfo): void {
-    if (!callback) {
-      throw new Error('[open purchase experience] Callback cannot be null');
+  export function openPurchaseExperience(callback: (error: SdkError | null) => void, planInfo?: PlanInfo): void;
+  /**
+   * @privateRemarks
+   * This function is the overloaded implementation of openPurchaseExperience.
+   * Since the method signatures of the v1 callback and v2 promise differ in the type of the first parameter,
+   * we need to do an extra check to know the typeof the @param1 to set the proper arguments of the utility function.
+   * @param param1
+   * @param param2
+   * @returns Promise that will be resolved when the operation has completed or rejected with SdkError value
+   */
+  export function openPurchaseExperience(
+    param1: ((error: SdkError | null) => void) | PlanInfo | undefined,
+    param2?: PlanInfo,
+  ): Promise<void> {
+    let callback: (error: SdkError | null) => void;
+    let planInfo: PlanInfo;
+    if (typeof param1 === 'function') {
+      callback = param1;
+      planInfo = param2;
+    } else {
+      planInfo = param1;
     }
+    const wrappedFunction: InputFunction<void> = () => {
+      return new Promise<void>(resolve => {
+        resolve(sendAndHandleSdkError('monetization.openPurchaseExperience', planInfo));
+      });
+    };
+
     ensureInitialized(FrameContexts.content);
-    sendMessageToParent('monetization.openPurchaseExperience', [planInfo], callback);
+    return callCallbackWithErrorOrResultOrNullFromPromiseAndReturnPromise(wrappedFunction, callback, planInfo);
```

Comment:
you're passing `planInfo` here but above in `wrappedFunction` you're not taking any parameters and instead close over `planInfo` directly. That seems weird? Should be either one or the other, but not both at the same time?

Reply:
I am not sure I understand your concern. 
Are you asking why I haven't passed the <planInfo> to <wrappedFunction>?If yes, 
Because the arrowFunction binds the context. 
And the wrappedFunction is not invoked during the declaration. It is invoked only once when it calls the utility function <callCallbackWithErrorOrResultOrNullFromPromiseAndReturnPromise>.

Reply:
3rd argument to `callCallbackWithErrorOrResultOrNullFromPromiseAndReturnPromise` is internally passed into the provided function (`wrappedFunction` in this case):

```
export function callCallbackWithErrorOrResultOrNullFromPromiseAndReturnPromise<T>(
  funcHelper: InputFunction<T>,
  callback?: ErrorResultNullCallback<T>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ...args: any[]
): Promise<T> {
  const p = funcHelper(...args);
```

In your case, the function you're passing in does not take any parameters (as you mentioned yourself `wrappedFunction` closes over it already). So why are we passing `planInfo` to `callCallbackWithErrorOrResultOrNullFromPromiseAndReturnPromise` at all?

Reply:
We should either not pass it in to `callCallbackWithErrorOrResultOrNullFromPromiseAndReturnPromise`, or we should not have `wrappedFunction` close over it and instead take it as a parameter.

Reply:
Got it. I will update the code.